### PR TITLE
[SPARK-11791] Fix flaky test in BatchedWriteAheadLogSuite

### PR DESCRIPTION
**Pull request by brkyvz**
stack trace of failure:

```
org.scalatest.exceptions.TestFailedDueToTimeoutException: The code passed to eventually never returned normally. Attempted 62 times over 1.006322071 seconds. Last failure message: 
Argument(s) are different! Wanted:
writeAheadLog.write(
    java.nio.HeapByteBuffer[pos=0 lim=124 cap=124],
    10
);
-> at org.apache.spark.streaming.util.BatchedWriteAheadLogSuite$$anonfun$23$$anonfun$apply$mcV$sp$15.apply(WriteAheadLogSuite.scala:518)
Actual invocation has different arguments:
writeAheadLog.write(
    java.nio.HeapByteBuffer[pos=0 lim=124 cap=124],
    10
);
-> at org.apache.spark.streaming.util.WriteAheadLogSuite$BlockingWriteAheadLog.write(WriteAheadLogSuite.scala:756)
```

I believe the issue was that due to a race condition, the ordering of the events could be messed up in the final ByteBuffer, therefore the comparison fails.

By adding eventually between the requests, we make sure the ordering is preserved. Note that in real life situations, the ordering across threads will not matter.

Another solution would be to implement a custom mockito matcher that sorts and then compares the results, but that kind of sounds like overkill to me. Let me know what you think @tdas @zsxwing

### CODE REVIEW

**Comment by zsxwing** [comment]
LGTM

**Comment by tdas** [comment]
Rather than enforcing ordering, isn it cleaner to capture the passed arguments to `wal.write`, deserialize and compare as sets? 
Its cleaner because the unit test should test whether all the data is actually written or not, even when submitted concurrently. 

Actually, you dont need to use custom mockito matcher (that would be ugly). ArgumentCaptors are much less ugly I think. See [link](http://stackoverflow.com/questions/5981605/can-mockito-capture-arguments-of-a-method-called-multiple-times).

**Comment by brkyvz** [comment]
I didn't know about ArgumentCaptors. My custom matcher was practically going to do the same thing. It might be cleaner and even less fragile hopefully

**Comment by brkyvz** [comment]
@tdas Addressed

File: streaming/src/test/scala/org/apache/spark/streaming/util/WriteAheadLogSuite.scala [hunk 1]
```diff
@@ -508,14 +509,17 @@ class BatchedWriteAheadLogSuite extends CommonWriteAheadLogTests(
     blockingWal.allowWrite()
 
     val buffer1 = wrapArrayArrayByte(Array(event1))
-    val buffer2 = wrapArrayArrayByte(Array(event2, event3, event4, event5))
+    val buffer2 = Set(event2, event3, event4, event5)
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/util/WriteAheadLogSuite.scala]: This isnt a buffer any more. Maybe not have `buffer2` at all. And then `buffer1` can be either removed or renamed to `buffer`.

**Comment by brkyvz** [comment]
retest this please

**Comment by tdas** [comment]
I am merging this to master and 1.6. Thanks @brkyvz for fixing it.

**Pull request closed without merging**